### DISCUSSION TITLE
Pre-populated WasmDatabase (IndexedDB)

### DISCUSSION THREAD

**Discussion by im-trisha** [simolus3/drift#2374] [Q&A] [answered]
As i need FTS5 support, the only way i can have it on web (As far as i've understood) is using WasmSqlite3 + IndexedDB.
I've managed to make it work, and it is working in fact pretty well, but as the title says, i must use a pre-populated database from the assets directory, while [this](https://github.com/simolus3/drift/discussions/2120?sort=top) is really useful for a normal web database, i can't manage to find a way to do it in my current setup:
```dart
DatabaseConnection connect({bool isInWebWorker = false}) {
  if (_useWorker && !isInWebWorker) {
    final worker = SharedWorker('shared_worker.dart.js');
    return DatabaseConnection.delayed(
      connectToRemoteAndInitialize(worker.port!.channel()),
    );
  } else {
    return DatabaseConnection.delayed(Future.sync(() async {
      // Experimental wasm support
      final response = await http.get(Uri.parse('sqlite3.wasm'));
      final fs = await IndexedDbFileSystem.open(dbName: 'application_name');
      final sqlite3 = await WasmSqlite3.load(
        response.bodyBytes,
        SqliteEnvironment(fileSystem: fs),
      );

      final databaseImpl = WasmDatabase(sqlite3: sqlite3, path: 'db.sqlite');
      return DatabaseConnection(databaseImpl);
    }));
  }
}
```

The only way i tought was writing manually the file using dart:html and window.indexedDB, but:

- I'm not completely sure on how to do it
- It seems like a really stupid way to do it, and i probably shouldn't

Any help would be appreciated, thanks!

**Comment by simolus3** [comment, marked as answer]
Similar to the approach you've linked, it's possible to use the `IndexedDbFileSystem` directly to write data. The API expects full file names (so absolute paths) in every operation, but otherwise you can pretty much directly write the file:

```dart
import 'package:path/path.dart' as p;

const path = 'db.sqlite';
final absolutePath = p.absolute(path);

if (!fs.exists(absolutePath)) {
 Uint8List myInitialDatabase = await loadDatabaseFileFromAsset();
  fs
    ..createFile(absolutePath)
    ..write(absolutePath, myInitialDatabase, 0);
  await fs.flush();
}

final databaseImpl = WasmDatabase(sqlite3: sqlite3, path: path);
```

**Reply by coogle** [reply]
I was looking at this code and I think the APIs have significantly changed since this answer was written? I have this exact same use-case where I want to load an SQLite database from an external source (e.g. asset or downloaded from a server) but I've been digging around the source for awhile now (both `drift` and `sqlite3`), and it seems like the above code isn't accurate anymore?

Could you please, if you have a moment, provide a bit more of a complete example here of how you are writing the bits from the database file to the VFS exactly and passing it into Drift?

**Reply by coogle** [reply]
Go easy on me because I'm 100% sure this is wrong, but after digging through the source for quite awhile and getting lost I feel like this is at least heading in the generally right direction with what I'm trying to accomplish. I don't really understand yet how to make this work with web workers, as for now I'm just trying to get it to based on what you described above..

- Load the DB from `assets/frontend.sqlite`
- Ultimately load the DB by downloading it from the server as needed
- Perform a simple `db.mytable().get()` call that returns the first 10 rows..

```dart
DatabaseConnection connect()
{
    return DatabaseConnection.delayed(Future(() async {

      final response = await http.get(Uri.parse('sqlite3.wasm'));

      IndexedDbFileSystem fs = await IndexedDbFileSystem.open(dbName: 'my_app');
      Uint8List database = (await rootBundle.load('assets/frontend.sqlite')).buffer.asUint8List();

      XOpenResult result = fs.xOpen(Sqlite3Filename(p.absolute('frontend.sqlite')), SqlFlag.SQLITE_OPEN_CREATE);

      result.file.xWrite(database, 0);

      await fs.flush();

      final db = await WasmDatabase(
          sqlite3: await WasmSqlite3.load(response.bodyBytes),
          path: p.absolute('frontend.sqlite'),
          fileSystem: fs
      );

      return DatabaseConnection(db);
    }));
}
```

While the below compiles, it throws an error `while opening the database, bad parameter or other API misuse, bad parameter or other API misuse` which seems to be coming from SQLite itself with a simple `SELECT * FROM foo LIMIT 10` query.

Also, when I was inspecting the IndexedDb I saw this, which doesn't seem right...Note the Exception (it's cut off)

![image](https://github.com/simolus3/drift/assets/1863471/27a5b6b8-da10-4234-b396-2a3e50fd3f99)

**Reply by simolus3** [reply]
With version 2 of the `sqlite3` package, the web APIs have changed a bit (as you've seen).

The good news is that drift is able to handle more of the required setup work internally now. You can use the [new web implementation](https://drift.simonbinder.eu/web/) with `WasmDatabase.open()`. That method has a `initializeDatabase` callback which can be used to load the initial asset when the app is first opened. Migrating from the old approach would require [a migration](https://drift.simonbinder.eu/web/#migrating-from-existing-web-databases) if loosing data is unacceptable to you.

The old way continues to be supported, but is more verbose now since it matches the sqlite3 APIs for a file system wrapper more closely:

```dart
DatabaseConnection connect() {
  return DatabaseConnection.delayed(Future(() async {
    IndexedDbFileSystem fs = await IndexedDbFileSystem.open(dbName: 'my_app');
    final path = p.absolute('frontend.sqlite');

    // Check if the database exists, and create it otherwise
    if (fs.xAccess(path, 0) == 0) {
      Uint8List database = (await rootBundle.load('assets/frontend.sqlite'))
          .buffer
          .asUint8List();

      final (file: file, outFlags: _) =
          fs.xOpen(Sqlite3Filename(path), SqlFlag.SQLITE_OPEN_CREATE);
      file.xWrite(database, 0);
      file.xClose();

      await fs.flush();
    }

    final sqlite3 = await WasmSqlite3.loadFromUrl(Uri.parse('sqlite3.wasm'));

    // The API misuse errors happen without this line
    sqlite3.registerVirtualFileSystem(fs, makeDefault: true);

    final db = WasmDatabase(
      sqlite3: sqlite3,
      path: path,
      fileSystem: fs,
    );

    return DatabaseConnection(db);
  }));
}
```

**Reply by coogle** [reply]
Thanks so much! Funny enough, not literally 5 minutes before you posted this answer I actually figured it out and my code looks unsurprisingly almost identical to yours. 

One thing I found that didn't seem to work here is your use of `p.absolute()`... I found that if I used `p.absolute()` I ended up with two databases: 

`/http://localhost:1234/frontend.sqlite` 
`http://localhost:1234/frontend.sqlite`

Instead, I had to just skip using the `path` package and pass `Sqlite3Filename('/frontend.sqlite')` to `xOpen` while passing just `frontend.sqlite` to `path` in `WasmDatabase`
